// @babylonjs/addons/atmosphere/atmospherePhysicalProperties.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { Observable } from "@babylonjs/core/Misc/observable.js";
import { Vector3 } from "@babylonjs/core/Maths/math.vector.js";
const DefaultPlanetRadius = 6360.0;
const DefaultPlanetRadiusOffset = 0.01;
const DefaultAtmosphereThickness = 100.0;
// The scattering and absorption values are per kilometer measured at sea level.
const DefaultPeakRayleighScattering = new Vector3(0.005802, 0.013558, 0.0331);
const DefaultPeakMieScattering = new Vector3(0.003996, 0.003996, 0.003996);
const DefaultPeakMieAbsorption = new Vector3(0.000444, 0.000444, 0.000444);
const DefaultPeakOzoneAbsorption = new Vector3(0.00065, 0.001881, 0.000085);
/**
 * Describes the physical properties of the atmosphere. Assumes a spherical planet.
 * - "radius" values describe a distance from the planet's center.
 * - "height" values describe a distance from the planet's surface.
 * - Distances are in kilometers unless otherwise specified. Angles are in radians.
 */
export class AtmospherePhysicalProperties {
    /**
     * The radius of the planet in kilometers.
     */
    get planetRadius() {
        return this._planetRadius;
    }
    set planetRadius(value) {
        if (this._planetRadius !== value) {
            this._planetRadius = value;
            this._recomputeDimensionalParameters();
            this.onChangedObservable.notifyObservers(this);
        }
    }
    /**
     * The squared radius of the planet in kilometers.
     */
    get planetRadiusSquared() {
        return this._planetRadiusSquared;
    }
    /**
     * Offset applied to view points near the planet's surface. This should be greater than 0.
     * It prevents rendering issues close to the planet's surface.
     */
    get planetRadiusOffset() {
        return this._planetRadiusOffset;
    }
    set planetRadiusOffset(value) {
        if (this._planetRadiusOffset !== value) {
            this._planetRadiusOffset = value;
            this._recomputeDimensionalParameters();
            this.onChangedObservable.notifyObservers(this);
        }
    }
    /**
     * This is the {@link planetRadius} with the additional {@link planetRadiusOffset}, in kilometers.
     */
    get planetRadiusWithOffset() {
        return this._planetRadiusWithOffset;
    }
    /**
     * The thickness of the atmosphere measured in kilometers.
     */
    get atmosphereThickness() {
        return this._atmosphereThickness;
    }
    set atmosphereThickness(value) {
        if (this._atmosphereThickness !== value) {
            this._atmosphereThickness = value;
            this._recomputeDimensionalParameters();
            this.onChangedObservable.notifyObservers(this);
        }
    }
    /**
     * The combined planet radius and atmosphere thickness in kilometers.
     */
    get atmosphereRadius() {
        return this._atmosphereRadius;
    }
    /**
     * The atmosphere radius squared in kilometers.
     */
    get atmosphereRadiusSquared() {
        return this._atmosphereRadiusSquared;
    }
    /**
     * Horizon distance from the planet's surface to the outer edge of the atmosphere in kilometers.
     */
    get horizonDistanceToAtmosphereEdge() {
        return this._horizonDistanceToAtmosphereEdge;
    }
    /**
     * Horizon distance from the planet's surface to the outer edge of the atmosphere, squared, in kilometers.
     */
    get horizonDistanceToAtmosphereEdgeSquared() {
        return this._horizonDistanceToAtmosphereEdgeSquared;
    }
    /**
     * The scale applied to {@link peakRayleighScattering}.
     */
    get rayleighScatteringScale() {
        return this._rayleighScatteringScale;
    }
    set rayleighScatteringScale(value) {
        if (this._rayleighScatteringScale !== value) {
            this._rayleighScatteringScale = value;
            this._recomputeRayleighScattering();
            this.onChangedObservable.notifyObservers(this);
        }
    }
    /**
     * The Rayleigh scattering per kilometer at sea level for red, green, and blue wavelengths.
     */
    get peakRayleighScattering() {
        return this._peakRayleighScattering;
    }
    set peakRayleighScattering(value) {
        if (!this._peakRayleighScattering.equals(value)) {
            this._peakRayleighScattering.copyFrom(value);
            this._recomputeRayleighScattering();
            this.onChangedObservable.notifyObservers(this);
        }
    }
    /**
     * The Rayleigh scattering per kilometer at sea level for red, green, and blue wavelengths.
     * This value cannot be set directly. It is inferred by scaling {@link peakRayleighScattering} by {@link rayleighScatteringScale}.
     */
    get rayleighScattering() {
        return this._rayleighScattering;
    }
    /**
     * The scale applied to {@link peakMieScattering}.
     */
    get mieScatteringScale() {
        return this._mieScatteringScale;
    }
    set mieScatteringScale(value) {
        if (this._mieScatteringScale !== value) {
            this._mieScatteringScale = value;
            this._recomputeMieScattering();
            this.onChangedObservable.notifyObservers(this);
        }
    }
    /**
     * The Mie scattering per kilometer at sea level for red, green, and blue wavelengths.
     */
    get peakMieScattering() {
        return this._peakMieScattering;
    }
    set peakMieScattering(value) {
        if (!this._peakMieScattering.equals(value)) {
            this._peakMieScattering.copyFrom(value);
            this._recomputeMieScattering();
            this.onChangedObservable.notifyObservers(this);
        }
    }
    /**
     * The Mie scattering per kilometer at sea level for red, green, and blue wavelengths.
     * This value cannot be set directly. It is inferred by scaling {@link mieScatteringScale} by {@link peakMieScattering}.
     */
    get mieScattering() {
        return this._mieScattering;
    }
    /**
     * The scale applied to {@link peakMieAbsorption}.
     */
    get mieAbsorptionScale() {
        return this._mieAbsorptionScale;
    }
    set mieAbsorptionScale(value) {
        if (this._mieAbsorptionScale !== value) {
            this._mieAbsorptionScale = value;
            this._recomputeMieAbsorption();
            this.onChangedObservable.notifyObservers(this);
        }
    }
    /**
     * The Mie absorption per kilometer at sea level for red, green, and blue wavelengths.
     */
    get peakMieAbsorption() {
        return this._peakMieAbsorption;
    }
    set peakMieAbsorption(value) {
        if (!this._peakMieAbsorption.equals(value)) {
            this._peakMieAbsorption.copyFrom(value);
            this._recomputeMieAbsorption();
            this.onChangedObservable.notifyObservers(this);
        }
    }
    /**
     * The Mie absorption per kilometer at sea level for red, green, and blue wavelengths.
     * This value cannot be set directly. It is inferred by scaling {@link mieAbsorptionScale} by {@link peakMieAbsorption}.
     */
    get mieAbsorption() {
        return this._mieAbsorption;
    }
    /**
     * The Mie extinction per kilometer at sea level for red, green, and blue wavelengths.
     * This value cannot be set directly. It is inferred by adding the {@link mieAbsorption} to the {@link mieScattering}.
     */
    get mieExtinction() {
        return this._mieExtinction;
    }
    /**
     * The scale applied to {@link peakOzoneAbsorption}.
     */
    get ozoneAbsorptionScale() {
        return this._ozoneAbsorptionScale;
    }
    set ozoneAbsorptionScale(value) {
        if (this._ozoneAbsorptionScale !== value) {
            this._ozoneAbsorptionScale = value;
            this._recomputeOzoneAbsorption();
            this.onChangedObservable.notifyObservers(this);
        }
    }
    /**
     * The ozone absorption per kilometer measured at a height corresponding to it's peak concentration,
     * for red, green, and blue wavelengths.
     */
    get peakOzoneAbsorption() {
        return this._peakOzoneAbsorption;
    }
    set peakOzoneAbsorption(value) {
        if (!this._peakOzoneAbsorption.equals(value)) {
            this._peakOzoneAbsorption.copyFrom(value);
            this._recomputeOzoneAbsorption();
            this.onChangedObservable.notifyObservers(this);
        }
    }
    /**
     * The ozone absorption per kilometer at sea level for red, green, and blue wavelengths.
     * This value cannot be set directly. It is inferred by scaling {@link peakOzoneAbsorption} by {@link ozoneAbsorptionScale}.
     */
    get ozoneAbsorption() {
        return this._ozoneAbsorption;
    }
    /**
     * Constructs the {@link AtmospherePhysicalProperties}.
     * @param options - The options for the {@link AtmospherePhysicalProperties}.
     */
    constructor(options) {
        /**
         * Notification for when properties of the {@link AtmospherePhysicalProperties} are changed.
         */
        this.onChangedObservable = new Observable();
        this._peakRayleighScattering = new Vector3();
        this._peakMieScattering = new Vector3();
        this._peakMieAbsorption = new Vector3();
        this._peakOzoneAbsorption = new Vector3();
        // Inferred values.
        this._planetRadiusWithOffset = 0;
        this._planetRadiusSquared = 0;
        this._atmosphereRadius = 0;
        this._atmosphereRadiusSquared = 0;
        this._horizonDistanceToAtmosphereEdge = 0;
        this._horizonDistanceToAtmosphereEdgeSquared = 0;
        this._rayleighScattering = new Vector3();
        this._mieScattering = new Vector3();
        this._mieAbsorption = new Vector3();
        this._mieExtinction = new Vector3();
        this._ozoneAbsorption = new Vector3();
        this._planetRadius = options?.planetRadius ?? DefaultPlanetRadius;
        this._planetRadiusOffset = options?.planetRadiusOffset ?? DefaultPlanetRadiusOffset;
        this._atmosphereThickness = options?.atmosphereThickness ?? DefaultAtmosphereThickness;
        this._rayleighScatteringScale = options?.rayleighScatteringScale ?? 1.0;
        this._peakRayleighScattering.copyFrom(options?.peakRayleighScattering ?? DefaultPeakRayleighScattering);
        this._mieScatteringScale = options?.mieScatteringScale ?? 1.0;
        this._peakMieScattering.copyFrom(options?.peakMieScattering ?? DefaultPeakMieScattering);
        this._mieAbsorptionScale = options?.mieAbsorptionScale ?? 1.0;
        this._peakMieAbsorption.copyFrom(options?.peakMieAbsorption ?? DefaultPeakMieAbsorption);
        this._ozoneAbsorptionScale = options?.ozoneAbsorptionScale ?? 1.0;
        this._peakOzoneAbsorption.copyFrom(options?.peakOzoneAbsorption ?? DefaultPeakOzoneAbsorption);
        // Compute inferred values.
        this._recomputeDimensionalParameters();
        this._recomputeRayleighScattering();
        this._recomputeMieScattering();
        this._recomputeMieAbsorption();
        this._recomputeOzoneAbsorption();
    }
    _recomputeDimensionalParameters() {
        this._planetRadiusWithOffset = this._planetRadius + this._planetRadiusOffset;
        this._planetRadiusSquared = this._planetRadius * this._planetRadius;
        this._atmosphereRadius = this._planetRadius + this._atmosphereThickness;
        this._atmosphereRadiusSquared = this._atmosphereRadius * this._atmosphereRadius;
        this._horizonDistanceToAtmosphereEdgeSquared = this._atmosphereRadiusSquared - this._planetRadiusSquared;
        this._horizonDistanceToAtmosphereEdge = Math.sqrt(this._horizonDistanceToAtmosphereEdgeSquared);
    }
    _recomputeRayleighScattering() {
        this._peakRayleighScattering.scaleToRef(this._rayleighScatteringScale, this._rayleighScattering);
    }
    _recomputeMieScattering() {
        this._peakMieScattering.scaleToRef(this._mieScatteringScale, this._mieScattering);
        this._recomputeMieExtinction();
    }
    _recomputeMieAbsorption() {
        this._peakMieAbsorption.scaleToRef(this._mieAbsorptionScale, this._mieAbsorption);
        this._recomputeMieExtinction();
    }
    _recomputeMieExtinction() {
        this._mieAbsorption.addToRef(this._mieScattering, this._mieExtinction);
    }
    _recomputeOzoneAbsorption() {
        this._peakOzoneAbsorption.scaleToRef(this._ozoneAbsorptionScale, this._ozoneAbsorption);
    }
}
